mmeds/util.py

```diff
@@ -1000,55 +1002,87 @@ def make_pheniqs_config(reads_forward, reads_reverse, barcodes_forward, barcodes
     return out_s
 
 
-def strip_error_barcodes(num_allowed_errors, map_hash, input_dir, output_dir):
-    # Strip errors for each fastq.gz file in input_dir
-    count = 1
-    for filename in os.listdir(input_dir):
-        sample = ''
-        # Match sample file to sample in hash
-        for key in map_hash:
-            if filename.startswith(key):
-                sample = key
-                break
-        if not sample == '':
-            # Open sample file for reading
-            f = gzip.open(Path(input_dir) / filename, mode='rt')
-            out = ''
-
-            line = f.readline()
-            # Compare each line's barcodes with the expected barcodes and count diff
-            while line is not None and not line == '':
-                code = line[len(line)-18:len(line)-1].split('-')
-                diff = 0
-                # Compare forward barcodes and count diff
-                for i in range(len(code[0])):
-                    if not code[0][i] == map_hash[sample][0][i]:
-                        diff += 1
-                # Compare reverse barcodes and count diff
-                for i in range(len(code[1])):
-                    if not code[1][i] == map_hash[sample][1][i]:
-                        diff += 1
-
-                # If diff does not exceed N, write read to output
-                if diff <= num_allowed_errors:
-                    out += line
-                    out += f.readline()
-                    out += f.readline()
-                    out += f.readline()
-                # Else skip read
-                else:
-                    f.readline()
-                    f.readline()
-                    f.readline()
+def strip_error_barcodes(num_allowed_errors, mapping_file, input_dir, verbose):
+    """
+    Strip reads with errors from demultiplexed fastq files and return new file content
+    =================================================================================
+    :num_allowed_errors: Maximum number of errors in barcode pairs to not be stripped
+    :mapping_file: File containing sample IDs and barcode pairs
+    :input_dir: Directory containing demultiplexed fastq files
+    :verbose: Print information to stdout
+    """
+    # Read in mapping file
+    output_content = {}
+    map_df = pd.read_csv(Path(mapping_file), sep='\t', header=[0, 1], na_filter=False)
+
+    # Only three columns are needed
+    sample_ids = map_df['#SampleID']['#q2:types']
+    forward_barcodes = map_df['BarcodeSequence']['categorical']
```

Comment:
store hard coded mapping file strings somewhere maybe

Reply:
Things like `#SampleID`, `BarcodeSquence` and `BarcodeSquenceR` should not be hardcoded but (ideally) imported or at least set as function parameters with default values.